lib/rate_limit_sink.c

```diff
@@ -0,0 +1,152 @@
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+#include <tlog/pkt.h>
+#include <tlog/sink.h>
+#include <tlog/grc.h>
+#include <tlog/rc.h>
+#include <tlog/rate_limit_sink.h>
+
+/** Rate Limit sink instance */
+struct tlog_rate_limit_sink {
+    struct tlog_sink        sink;           /**< Abstract sink instance */
+    struct tlog_sink       *dsink;          /**< "Destination" (logging) sink */
+    bool                    dsink_owned;    /**< True if dsink is owned */
+    uint64_t                rate;           /**< Cutoff rate */
+
+    const struct tlog_pkt       *prev_pkt;  /**< Previous packet to check */
+    struct tlog_pkt_pos   *prev_ppos;       /**< Previous position for writing */
+    const struct tlog_pkt_pos   *prev_end;  /**< Previous end pos for writing */
+
+    struct timespec        prev_ts;       /**< Previous timestamp for time elapsed */
+};
+
+static bool
+tlog_rate_limit_sink_is_valid(const struct tlog_sink *sink)
+{
+    struct tlog_rate_limit_sink *rate_limit_sink =
+                                (struct tlog_rate_limit_sink *)sink;
+    return rate_limit_sink != NULL &&
+           tlog_sink_is_valid(rate_limit_sink->dsink);
+}
+
+static void
+tlog_rate_limit_sink_cleanup(struct tlog_sink *sink)
+{
+    struct tlog_rate_limit_sink *rate_limit_sink =
+                                (struct tlog_rate_limit_sink *)sink;
+    assert(rate_limit_sink != NULL);
+    if(rate_limit_sink->dsink_owned == true){
+        tlog_sink_destroy(rate_limit_sink->dsink);
+        rate_limit_sink->dsink_owned = false;
+    }
+    rate_limit_sink->prev_pkt = NULL;
+    rate_limit_sink->prev_ppos = NULL;
+    rate_limit_sink->prev_end = NULL;
+    (void) rate_limit_sink; /* unused unless asserts are enabled */
+}
+
+static tlog_grc
+tlog_rate_limit_sink_init(struct tlog_sink *sink, va_list ap)
+{
+    struct tlog_rate_limit_sink *rate_limit_sink =
+                                 (struct tlog_rate_limit_sink *)sink;
+    struct tlog_sink *pdsink = va_arg(ap, struct tlog_sink *);
+
+    if(pdsink == NULL){
+      goto error;
+    }
+    rate_limit_sink->dsink = pdsink;
+    rate_limit_sink->dsink_owned = (bool)va_arg(ap, int);
+    rate_limit_sink->rate = va_arg(ap, uint64_t);
+    rate_limit_sink->prev_pkt = NULL;
+    rate_limit_sink->prev_ppos = NULL;
+    rate_limit_sink->prev_end = NULL;
+
+    return TLOG_RC_OK;
+
+error:
+    tlog_rate_limit_sink_cleanup(sink);
+    return TLOG_GRC_ERRNO;
+}
+
+static tlog_grc
+tlog_rate_limit_sink_write(struct tlog_sink *sink,
+        const struct tlog_pkt *pkt,
+        struct tlog_pkt_pos *ppos,
+        const struct tlog_pkt_pos *end)
+{
+    struct tlog_rate_limit_sink *rate_limit_sink =
+        (struct tlog_rate_limit_sink *)sink;
+    tlog_grc grc;
+
+    if(rate_limit_sink->prev_pkt == NULL) {
+        grc = TLOG_RC_OK;
+        goto end;
+    }
+    if(rate_limit_sink->prev_pkt->type != TLOG_PKT_TYPE_IO || rate_limit_sink->rate == 0){
```

Comment:
My understanding could be wrong, but I didn't think it would bypass rate-limiting.  Technically, in a sense, I suppose it would since I'm not checking rate on window resizes. However, the focus was limiting data from I/O, not window resizing, so window resize rates wouldn't matter (if I'm understanding your comment/question correctly).  I also can't think of a way someone would be able to bypass rate-limiting for I/O data through window resizing since the packets seem to be sequential and not parallel.

Is that along the lines you were thinking?